### ISSUE TITLE
Make a JShell kernel

### ISSUE THREAD

**Issue by dsblank** [twosigma/beakerx#5428]
Java9 is about ready to release its RC1, with the final version due out in about a month (July 2017):

http://openjdk.java.net/projects/jdk9/

It would be fairly easy to make a beakerx JShell-based kernel:

```java
import jdk.jshell.*;

public class JShellKernel {
    JShell js;
    JShellKernel() {
	js = JShell.create();
    }

    public void eval(String code) {
	js.eval(code);
        // etc.
    }
}
```

There are differences between a JShellKernel and the current beakerx JavaKernel (this is to the best of my knowledge, not being an expert in JShell nor Beakerx JavaKernel):

1. JShell can evaluate snippets and expressions (eg, "1 + 1"). See http://download.java.net/java/jdk9/docs/api/jdk.jshell-summary.html. JavaKernel requires each cell to be a complete set of Java statement(s).
2. JShell has optional semicolons; JavaKernel does not.
3. JShell doesn't support return as a snippet; JavaKernel does support return.
4. JShell allows variables to persist between evals; JavaKernel variables live only for the life of the cell.
5. JShell allows "forward reference": code may reference items not defined yet. This allows writing code that has co-dependencies. I don't think this type of code development is possible in JavaKernel.

From a pedagogical standpoint, none of these differences may be desired, and the JavaKernel may be prefered in the classroom. For example, if students learn that they don't need semicolons, then there is some cognitive overhead when transitioning to full Java.

On a related note, parts of Java9 JShell might be usable in the JavaKernel, if you switch to java9. For example, JShell also has functionality for code-completion.

**Comment by scottdraves** [comment]
Yup, agreed, when we switch to java9.  We had an issue for this in classic Beaker: https://github.com/twosigma/beakerx/issues/3585
That switch might be a long way away however.

**Comment by pdeitel** [comment]
Hi All, Curious if any work has been done on incorporating Java 9 support in BeakerX yet. Thanks.

**Comment by scottdraves** [comment]
No work has been done but we are closer now that java9 is out.
Normally we get the jdk from conda, and java9 is not there yet afaik:
https://github.com/conda-forge/openjdk-feedstock/issues/21
We could do the development without that though.
I think best would be to support it as an additional JShell kernel, and leave our Java kernel as it is.

**Comment by dsblank** [comment]
I agree: a JShell Kernel would be a different beast, and the choice of which to use should be left to the user. There are pros and cons to both. (I really like the idea of a JShell kernel... our early prototype [1] can be quite useful, and I see some other new java9 kernel projects. But building on the Beaker infrastructure would be great!) 

[1] - https://jupyter.brynmawr.edu/services/public/dblank/Experiments/Java9.ipynb

**Comment by pdeitel** [comment]
Just checking whether any more work has been done on Java 9 JShell kernel integration. If so, I’d love to play around with it and provide feedback.

Also, do you know of a Jupyter kernel that fully supports Java 9?

Best,
Paul



On Nov 8, 2017, at 11:21 AM, Doug Blank <notifications@github.com> wrote:

I agree: a JShell Kernel would be a different beast, and the choice of which to use should be left to the user. There are pros and cons to both. (I really like the idea of a JShell kernel... our early prototype [1] can be quite useful, and I see some other new java9 kernel projects. But building on the Beaker infrastructure would be great!)

[1] - https://jupyter.brynmawr.edu/services/public/dblank/Experiments/Java9.ipynb <https://jupyter.brynmawr.edu/services/public/dblank/Experiments/Java9.ipynb>

**Comment by scottdraves** [comment]
We are blocked on conda supporting java9. See the above feedstock issue.

**Comment by pdeitel** [comment]
That’s what I thought. And no feedback from them on when they might add Java 9?



On Feb 11, 2018, at 11:14 AM, Scott Draves <notifications@github.com> wrote:

We are blocked on conda supporting java9. See the above feedstock issue.

**Comment by scottdraves** [comment]
Correct. Maybe voting for the issue will help?

**Comment by benmccann** [comment]
I'm curious if this is something that might be prioritized soon since Conda has added Java 9 support now

**Comment by benmccann** [comment]
https://github.com/SpencerPark/IJava uses JShell and is MIT licensed in case it is helpful to have an example of another JShell kernel
